plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,308 @@
+      parameterTypeStringLength = fromIntegral $ T.length parameterTypeString
+      start = totalPrefixLength + prefixLength
+   in (start, start + parameterTypeStringLength)
+        : findParameterStringRanges
+          (totalPrefixLength + prefixLength + parameterTypeStringLength)
+          (T.drop (fromIntegral parameterTypeStringLength) match)
+          restParameterTypeStrings
+
+-- similar to 'splitFunTys' but
+--   1) the result (last) type is included and
+--   2) toplevel foralls are ignored
+splitFunTysIgnoringForAll :: Type -> [(Type, Maybe FunTyFlag)]
+splitFunTysIgnoringForAll ty = case ty & dropForAlls & splitFunTy_maybe of
+  Just (funTyFlag, _mult, parameterType, resultType) ->
+    (parameterType, Just funTyFlag) : splitFunTysIgnoringForAll resultType
+  Nothing -> [(ty, Nothing)]
+
+notTypeConstraint :: (Type, Maybe FunTyFlag) -> Bool
+notTypeConstraint (_type, Just FTF_T_T) = True
+notTypeConstraint (_type, Nothing)      = True
+notTypeConstraint _                     = False
+
+extractInfoFromSmallestContainingFunctionApplicationAst ::
+  Position -> HieASTs a -> (RealSrcSpan -> HieAST a -> Maybe b) -> [b]
+extractInfoFromSmallestContainingFunctionApplicationAst position hieAsts extractInfo =
+  M.elems $ flip M.mapMaybeWithKey (getAsts hieAsts) $ \hiePath hieAst ->
+    smallestContainingSatisfying (positionToSpan hiePath position) (nodeHasAnnotation ("HsApp", "HsExpr")) hieAst
+      >>= extractInfo (positionToSpan hiePath position)
+  where
+    positionToSpan hiePath position =
+      let loc = mkLoc hiePath position in mkRealSrcSpan loc loc
+    mkLoc (LexicalFastString hiePath) (Position line character) =
+      mkRealSrcLoc hiePath (fromIntegral line + 1) (fromIntegral character + 1)
+
+type Annotation = (FastStringCompat, FastStringCompat)
+
+nodeHasAnnotation :: Annotation -> HieAST a -> Bool
+nodeHasAnnotation annotation hieAst = case sourceNodeInfo hieAst of
+  Nothing       -> False
+  Just nodeInfo -> isAnnotationInNodeInfo annotation nodeInfo
+
+getLeftMostNode :: HieAST a -> HieAST a
+getLeftMostNode thisNode =
+  case nodeChildren thisNode of
+    []            -> thisNode
+    leftChild : _ -> getLeftMostNode leftChild
+
+getNodeNameAndTypes :: HieKind a -> HieAST a -> Maybe (Name, [Type])
+getNodeNameAndTypes hieKind hieAst =
+  if nodeHasAnnotation ("HsVar", "HsExpr") hieAst
+    then case hieAst & getSourceNodeIds & M.filter isUse & M.assocs of
+      [(identifier, identifierDetails)] ->
+        case extractName identifier of
+          Nothing -> Nothing
+          Just name ->
+            let mTypeOfName = identType identifierDetails
+                typesOfNode = case sourceNodeInfo hieAst of
+                  Nothing       -> []
+                  Just nodeInfo -> nodeType nodeInfo
+                allTypes = case mTypeOfName of
+                  Nothing -> typesOfNode
+                  Just typeOfName -> typeOfName : filter (isDifferentType typeOfName) typesOfNode
```

Comment:
Why do we need to filter here? worth explaining!

Reply:
The last type of `typesOfNode` may (always?) be the same of `typeOfName`.  To avoid generating two identical signature helps, we do a filtering here.

Do you think it is worth adding a comment for it?

Reply:
> I did not do so because I am not sure 1) if there always are duplicate types and 2) if the duplicate type is the last one of typesOfNode. If the answers of both questions are true, then we can use a more efficient implementation, i.e., dropEnd1, than filtering.

I am not 100% sure, but we can try it out, and if a test fails, then we know whether it is the same :)
Otherwise, someone will report a bug at some point in which case we can write a regression test and fix and document it for all enternity!